src/NuGet.Clients/VsExtension/VisualStudioAccountProvider.cs

```diff
@@ -0,0 +1,390 @@
+                    {
+                        // The account didn't have access and we are on a retry so the token didn't expire.
+                        // we either need to prompt the user for different creds or fail in this case
+                        ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                }
+            }
+            else if (posibleAccounts.Count > 1)
+            {
+                var accountsWithAccess = new List<AccountWithCreds>();
+                foreach (var account in posibleAccounts)
+                {
+                    var cred = await GetTokenFromAccount(
+                        account,
+                        provider,
+                        nonInteractive: true,
+                        cancellationToken:cancellationToken)
+                        .ConfigureAwait(false);
+
+                    var hasAccess = await AccountHasAccess(uri, proxy, cred, cancellationToken)
+                        .ConfigureAwait(false);
+                    if (hasAccess)
+                    {
+                        accountsWithAccess.Add(new AccountWithCreds(account, cred));
+                    }
+                }
+
+                if (accountsWithAccess.Count == 1)
+                {
+                    ret = accountsWithAccess[0].Creds;
+                }
+                else
+                {
+                    // we couldn't finde a unique account with access to the endpoint so we are going to have
+                    // to ask the user...
+                    ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                        .ConfigureAwait(false);
+                }
+
+            }
+            else // count == 0 so we should prompt the user
+            {
+                ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            if (ret == null)
+            {
+                // No credentials found but we know that this is a VSO endpoint so we want to throw an
+                // exception to prevent the other providers from being called
+                throw new Exception("No valid credentials found for VSO account");
+            }
+
+            return ret;
+        }
+
+        // Logic shows UI and interacts with all mocked methods.  Mocking this as well.
+        internal virtual async Task<ICredentials> PromptUserForAccount(
+        	string tenentId, VSAccountProvider provider, bool nonInteractive, CancellationToken cancellationToken)
+        {
+            ICredentials ret = null;
+            if (nonInteractive)
+            {
+                //  If we are not supposed to interact with the user then we can't prompt for account so we
+                // need to fail.
+                return null;
```

Comment:
Seems we should throw here. null means the provider doesn't support the endpoint, and nuget will try other providers.